types/smtp-server/index.d.ts

```diff
@@ -1,386 +1,317 @@
+    maxClients?: number;
     /**
-     * boolean, if set to true expects to be behind a proxy that emits a 
+     * boolean, if set to true expects to be behind a proxy that emits a
      * PROXY header{http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt} (version 1 only)
-     * 
-     * @type {boolean}
-     * @memberOf SMTPServerOptions
      */
     useProxy?: boolean;
     /**
-     * boolean, if set to true, enables usage of 
-     * XCLIENT{http://www.postfix.org/XCLIENT_README.html} extension to override connection properties. 
+     * boolean, if set to true, enables usage of
+     * XCLIENT{http://www.postfix.org/XCLIENT_README.html} extension to override connection properties.
      * See session.xClient (Map object) for the details provided by the client
-     * 
-     * @type {boolean}
-     * @memberOf SMTPServerOptions
      */
     useXClient?: boolean;
     /**
-     * boolean, if set to true, enables usage of XFORWARD{http://www.postfix.org/XFORWARD_README.html} extension. 
+     * boolean, if set to true, enables usage of XFORWARD{http://www.postfix.org/XFORWARD_README.html} extension.
      * See session.xForward (Map object) for the details provided by the client
-     * 
-     * @type {boolean}
-     * @memberOf SMTPServerOptions
      */
     useXForward?: boolean;
     /**
      * boolean, if set to true use LMTP protocol instead of SMTP
-     * 
-     * @type {boolean}
-     * @memberOf SMTPServerOptions
      */
     lmtp?: boolean;
     /**
      * How many milliseconds of inactivity to allow before disconnecting the client (defaults to 1 minute)
-     * 
-     * @type {number}
-     * @memberOf SMTPServerOptions
      */
-    socketTimeout?: number;//millisceonds
+    socketTimeout?: ms;
     /**
-     * How many millisceonds to wait before disconnecting pending 
+     * How many millisceonds to wait before disconnecting pending
      * connections once server.close() has been called (defaults to 30 seconds)
-     * 
-     * @type {number}
-     * @memberOf SMTPServerOptions
      */
-    closeTimeout?: number;//millisceonds
+    closeTimeout?: ms;
     /**
      * The callback to handle authentications (see details https://github.com/andris9/smtp-server#handling-authentication)
-     * 
-     * 
-     * @memberOf SMTPServerOptions
      */
-    onAuth?: (auth: Authentication, session: Session, callback: (err?: Error, response?: AuthenticationResponse) => any) => any;
+    onAuth?(auth: SMTPServerAuthentication, session: SMTPServerSession, callback: (err: Error | null, response: SMTPServerAuthenticationResponse) => void): void;
     /**
      * The callback to handle the client connection. (see details https://github.com/andris9/smtp-server#validating-client-connection)
-     * 
-     * 
-     * @memberOf SMTPServerOptions
      */
-    onConnect?: (session: Session, callback: (err?: Error) => any) => any;
+    onConnect?(session: SMTPServerSession, callback: (err: Error | null) => void): void;
```

Comment:
First, thanks for a good work on the library typings update!

Why `callback` function type has been changed from `callback: (err?: Error) => any` to `callback: (err: Error | null) => void`. What is the advantage of changing optional argument to a composite `null` type? I can't find this in library documentation or in the `smtp-server` code...

Same question for other functions that have previously accepted optional arguments but were changed to `sometype | null` instead.

Reply:
I followed http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html so I used `void` rather than `any`.

> Don’t use the return type any for callbacks whose value will be ignored

The same for `err` argument: `null` is a convention for error-first callbacks. There is no reason to not follow this convention if an error is an only argument. And recommendation says:

> Do write callback parameters as non-optional

Reply:
From "Do's and Dont's":

> This has a very specific meaning: the done callback might be invoked with 1 argument or might be invoked with 2 arguments.

I think that's exactly our case for `onConnect`: implementor of `onConnect` can call this callback either with `Error` or without (just calling `callback()`). If we specify `err` argument as required composite `null` type, it will force `onConnect` implementations to always invoke callback with `callback(null)`. 

This is OK, but I don't understand what implementor will gain from this approach. Downside is that some developers use `no-null-keyword` setting in TSLint, and they'll have to silence `Use undefined instead of null` warnings for their `onConnect` implementations.

![](https://i.imgur.com/FEqk0Oc.jpg)

Same for other event-like functions: `onMailFrom`, `onRcptTo`, `onData`, `onClose`.

Anyway, that's a nitpick I'll approve your PR and if someone doesn't like this change, they'll fix it in the future.

Reply:
@andy-ms do you have some advice about best practices for callbacks?

Reply:
The two have different meanings: `(err?: Error)` may be `undefined` but not `null` and `err: Error | null` may be `null` but not `undefined`.
If either can happen you should use `err?: Error | null` as the best of both worlds.